Create new beatmap relationships

hburn7 opened the pull request:
Closes #582 
Closes #586
Closes #588 

This PR clears and resets all migrations. Care has been taken to ensure smooth deployment with no data loss.

hburn7 commented on DataWorkerService/Services/Implementations/OsuApiDataParserService.cs:
```diff
@@ -263,39 +263,65 @@ public IEnumerable<GameScore> CreateScores(IEnumerable<ApiGameScore> apiScores)
+        beatmap.BeatmapSetId = beatmapSet.Id;
+    }
+
+    private static BeatmapRankedStatus StatusFromString(string status)
+    {
+        return status.ToLower() switch
+        {
+            "graveyard" => BeatmapRankedStatus.Graveyard,
+            "wip" => BeatmapRankedStatus.WorkInProgress,
+            "pending" => BeatmapRankedStatus.Pending,
+            "ranked" => BeatmapRankedStatus.Ranked,
+            "approved" => BeatmapRankedStatus.Approved,
+            "qualified" => BeatmapRankedStatus.Qualified,
+            "loved" => BeatmapRankedStatus.Loved,
+            _ => throw new Exception($"Could not determine BeatmapRankedStatus from string {status}")
+        };
     }
```
Okay, this was added because I saw a string property called Status and didn't know how else to parse it, wasn't sure.

hburn7 commented on Database/Entities/Beatmap.cs:
```diff
@@ -1,20 +1,24 @@
 
 /// <summary>
-/// An osu! beatmap
+/// Represents a beatmap in the osu! API.
```
Yeah this was a bad change lol

hburn7 commented on Database/Entities/BeatmapAttributes.cs:
```diff
@@ -0,0 +1,37 @@
+[UsedImplicitly]
+public class BeatmapAttributes : EntityBase
+{
+    /// <summary>
+    /// Id of the associated beatmap
+    /// </summary>
+    [Column("beatmap_id")]
+    public int BeatmapId { get; set; }
+
+    /// <summary>
+    /// Mods applied to the beatmap
+    /// </summary>
+    [Column("mods")]
+    public Mods Mods { get; set; }
+
+    /// <summary>
+    /// Star rating of the beatmap with the applied mods
+    /// </summary>
+    [Column("sr")]
+    public double Sr { get; set; }
+
+    /// <summary>
+    /// The associated beatmap
+    /// </summary>
+    public Beatmap Beatmap { get; set; } = null!;
```
May I ask what the reason is for placing them at the bottom? I put this one at the top because all of the IDs would be next to each other. If we add future relationships, the new ID would go at the end of the table (bottom of class) so maybe bottom is best.

myssto replied:
> I put this one at the top because all of the IDs would be next to each other

Column order doesn't really exist with postgres if that is what you're referring to. And that relation convention is pretty standard for ef https://learn.microsoft.com/en-us/ef/core/modeling/relationships/conventions#discovering-foreign-key-properties

hburn7 replied:
I don't see anything in that article which defines a standard for where within the class the Id fields to be. The example classes are small enough to where the primary key is above the foreign key, which makes sense, but that's all I see. 

The columns are created in order when we have a migration which creates the table. So I see it as more convenient to have the FK Ids be next to each other in the class file now that we have the opportunity to recreate this table.

myssto replied:
Again the concept of concept of column order doesn't do anything functional in postgres. If you would like for the keys to be visible near each other when viewing the db I'm sure you can visually re-arrange them with DataGrip? You're right, it may not be based off of strictly defined convention (I swear I read that somewhere and I thought it would be in that doc) but what I outlined is the way it's been done consistently with every other entity in the project. I would rather not make that exception (or change every one of them) just because it would look nicer when viewing the raw database...

hburn7 replied:
Okay, because it's an existing convention I'm okay with the change. This is getting into subjective entity design territory anyway, the main thing that matters is consistency.

myssto replied:
Agreed 👍

myssto replied:
This change still needs to be applied

hburn7 commented on Database/OtrContext.cs:
```diff
@@ -83,13 +85,81 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+            entity
+                .HasMany(b => b.Creators)
+                .WithMany(p => p.CreatedBeatmaps)
+                .UsingEntity<Dictionary<string, object>>(
+                    "__join__beatmap_creators",
+                    r => r.HasOne<Player>()
+                        .WithMany()
+                        .HasForeignKey("player_id")
+                        .HasConstraintName("FK___join__beatmap_creators_Player"),
+                    l => l.HasOne<Beatmap>()
+                        .WithMany()
+                        .HasForeignKey("beatmap_id")
+                        .HasConstraintName("FK___join__beatmap_creators_Beatmap"));
```
I believe this is how the other join table is created. For many-to-many relationships you have to create a join table, this is a way to do that without creating an entity. EF manages the entity and relationships under the hood this way.

hburn7 replied:
See here:

```csharp
entity
                .HasMany(t => t.PooledBeatmaps)
                .WithMany(pb => pb.TournamentsPooledIn)
                .UsingEntity<Dictionary<string, object>>(
                    "__join__pooled_beatmaps",
                    r => r.HasOne<Beatmap>()
                        .WithMany()
                        .HasForeignKey("beatmap_id")
                        .HasConstraintName("FK___join__pooled_beatmaps_Beatmap"),
                    l => l.HasOne<Tournament>()
                        .WithMany()
                        .HasForeignKey("tournament_id")
                        .HasConstraintName("FK___join__pooled_beatmaps_Tournament"));
```

myssto replied:
I'm 99% certain that you can just remove everything below `.WithMany(...)` and it will let EF handle all of the convention. See https://learn.microsoft.com/en-us/ef/core/modeling/relationships/many-to-many#basic-many-to-many, where the example shows a similar setup, and says this as well:

> This relationship is [mapped by convention](https://learn.microsoft.com/en-us/ef/core/modeling/relationships/conventions). Even though it is not needed, an equivalent explicit configuration for this relationship is shown below as a learning tool:

```cs
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<Post>()
        .HasMany(e => e.Tags)
        .WithMany(e => e.Posts);
}
```

Its fine if you cba to change this, I just think we should avoid it in the future because by doing this we are tying the FK constraint names to string literals meant to represent `[Column]` attributes elsewhere.

hburn7 replied:
Huh, yeah you're right. Okay I will edit this and fix the other implementation.

hburn7 commented on DataWorkerService/Services/Implementations/OsuApiDataParserService.cs:
```diff
@@ -263,39 +263,65 @@ public IEnumerable<GameScore> CreateScores(IEnumerable<ApiGameScore> apiScores)
-        beatmap.Title = apiBeatmap.Beatmapset.Title;
-        beatmap.MapperName = apiBeatmap.Beatmapset.Creator ?? string.Empty;
-        beatmap.MapperId = apiBeatmap.Beatmapset.CreatorId ?? default;
+        Player creatorPlayer = await playersRepository.GetOrCreateAsync(apiBeatmap.Beatmapset.CreatorId.Value);
```
Okay. I'm thinking we can make MatchUser inherit from User outside of the Country object. 

![image](https://github.com/user-attachments/assets/95901c78-5409-4821-8457-223fdb56651a)

hburn7 commented on DataWorkerService/Services/Interfaces/IOsuApiDataParserService.cs:
```diff
@@ -12,11 +14,13 @@ public interface IOsuApiDataParserService
     /// <param name="apiMatch"><see cref="MultiplayerMatch"/> data</param>
     Task ParseMatchAsync(Match match, MultiplayerMatch apiMatch);
+
+    Task ParseBeatmapAsync(Beatmap beatmap, BeatmapExtended apiBeatmap);
```
The previous implementation was `public static`, interfacing gives the access back to `TournamentDataProcessor.cs`.

hburn7 commented:
Pushed some changes:
- MatchUser now inherits from User (same for the json model).
- Addressed most code review comments

Still todo:
- Need to fix the database schemas and regenerate migrations
- Test everything

myssto commented on DataWorkerService/Services/Implementations/OsuApiDataParserService.cs:
```diff
@@ -263,25 +264,41 @@ public IEnumerable<GameScore> CreateScores(IEnumerable<ApiGameScore> apiScores)
+
+
+        // Create or update BeatmapSet
+        var beatmapSet = new BeatmapSet
+        {
+            OsuId = apiBeatmap.Beatmapset.Id,
+            Artist = apiBeatmap.Beatmapset.Artist,
+            Title = apiBeatmap.Beatmapset.Title,
+            CreatorId = creatorPlayer.Id,
+            RankedStatus = apiBeatmap.Beatmapset.RankedStatus,
+            SubmittedDate = apiBeatmap.Beatmapset.SubmittedDate,
+            RankedDate = apiBeatmap.Beatmapset.RankedDate
+        };
+
+        beatmap.BeatmapSet = beatmapSet;
+        beatmap.BeatmapSetId = beatmapSet.Id;
```
```suggestion
        beatmap.BeatmapSet = new BeatmapSet
        {
            OsuId = apiBeatmap.Beatmapset.Id,
            Artist = apiBeatmap.Beatmapset.Artist,
            Title = apiBeatmap.Beatmapset.Title,
            CreatorId = creatorPlayer.Id,
            RankedStatus = apiBeatmap.Beatmapset.RankedStatus,
            SubmittedDate = apiBeatmap.Beatmapset.SubmittedDate,
            RankedDate = apiBeatmap.Beatmapset.RankedDate
        };
```
`BeatmapSet.Id` is not available at this point. Better to create the entity by proxy and let EF assign the Id. Your code likely works fine but this is cleaner imo

myssto commented on Database/Entities/Beatmap.cs:
```diff
@@ -1,20 +1,24 @@
 using System.ComponentModel.DataAnnotations.Schema;
-using System.Diagnostics.CodeAnalysis;
 using Database.Enums;
+// ReSharper disable CollectionNeverUpdated.Global
```
Why not just use the suppress attribute instead?

myssto commented on API/DTOs/BeatmapSetCompactDTO.cs:
```diff
@@ -0,0 +1,57 @@
+    /// <summary>
+    /// The set creator
+    /// </summary>
+    public PlayerCompactDTO Creator { get; set; } = new();
```
```suggestion
    public PlayerCompactDTO? Creator { get; set; }
```
Bad! This is how we get gems like:
```json
"user": {
    "avatar_url": null,
    "country_code": null,
    "default_group": "default",
    "id": null,
    "is_active": false,
    "is_bot": false,
    "is_deleted": true,
    "is_online": false,
    "is_supporter": false,
    "last_visit": null,
    "pm_friends_only": false,
    "profile_colour": null,
    "username": "[deleted user]"
}
```

myssto commented on API/DTOs/BeatmapSetCompactDTO.cs:
```diff
@@ -0,0 +1,57 @@
+    /// <summary>
+    /// Id of the Player who created the set
+    /// </summary>
+    public int CreatorId { get; set; }
```
Move above `PlayerCompactDTO? Creator`

myssto commented on OsuApiClient/OsuClient.cs:
```diff
@@ -352,6 +352,24 @@ public void Dispose()
+                Credentials = _credentials,
+                Method = HttpMethod.Get,
+                Route = new Uri(Endpoints.Osu.Beatmaps, UriKind.Relative),
+                QueryParameters = beatmapIds.Select((id, index) => new KeyValuePair<string, string>($"ids%5B{index}%5D", id.ToString())).ToDictionary()
```
```suggestion
                QueryParameters = beatmapIds
                    .Unique()
                    .Select((id, index) => new KeyValuePair<string, string>($"ids%5B{index}%5D", id.ToString()))
                    .ToDictionary()
```

myssto commented on Database/Entities/Beatmap.cs:
```diff
@@ -6,15 +6,20 @@
+public class Beatmap : UpdateableEntityBase
 {
     /// <summary>
-    /// osu! id of the beatmap
+    /// Id of the associated beatmapset
+    /// </summary>
+    [Column("beatmapset_id")]
+    public int? BeatmapSetId { get; set; }
```
Please move this above `BeatmapSet`

myssto commented on Database/OtrContext.cs:
```diff
@@ -83,13 +85,72 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+            entity
+                .HasMany(b => b.Creators)
+                .WithMany(p => p.CreatedBeatmaps)
+                .UsingEntity("__join__beatmap_creators");
```
if this will be a pita to fix then dont, but I don't understand why we need to do this at all

hburn7 replied:
Because I would like for it to be abundantly clear that this is a join table and is not like other tables. This is something I don't want to change.

myssto commented on Database/OtrContext.cs:
```diff
@@ -753,16 +814,7 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
-                        .WithMany()
-                        .HasForeignKey("tournament_id")
-                        .HasConstraintName("FK_JoinTable_Tournament"));
+                .UsingEntity("__join__pooled_beatmaps");
```
Same with this

myssto reviewed (approved):
Really great work, just some final small stuff and you can merge when youre ready
